### ISSUE TITLE
syntastic & lightline-whitespace won't show up on statusbar

### ISSUE THREAD

**Issue by ondiversion** [itchyny/lightline.vim#616]
Hi,

I am posting here to look for some help regarding implemeting 'syntastic & lightline whitespace' into my statusbar.
I, at least, tried my best to follow exact instructions given by each plugins manual,
and searched and looked up for others' .vimrc configurations.
But, still, it is not working for me, so I finally came here to ask for some advise.

Below is my .vimrc configuration that is realted to 'lightline.vim, vim-lightline-whitespace, and syntastic':

1. Plugin list

```vimrc
"------------------------------------------------------------------------
                 " Vim Plug
"------------------------------------------------------------------------

call plug#begin('~/.vim/plugged')


"utility
Plug 'junegunn/fzf', { 'do': { -> fzf#install() } }
Plug 'junegunn/fzf.vim'
Plug 'mileszs/ack.vim'
Plug 'kien/ctrlp.vim'
Plug 'vifm/vifm.vim'
Plug 'tpope/vim-fugitive'
Plug 'preservim/nerdtree'

"wordprocessing
Plug 'vim-pandoc/vim-pandoc'
Plug 'vim-pandoc/vim-pandoc-syntax'
Plug 'vim-pandoc/vim-rmarkdown'
Plug 'plasticboy/vim-markdown'
Plug 'scrooloose/syntastic'
Plug 'tpope/vim-surround'
Plug 'scrooloose/nerdcommenter'
Plug 'nathanaelkane/vim-indent-guides'
Plug 'deponian/vim-lightline-whitespace'
Plug 'junegunn/goyo.vim'

"accessibilities
Plug 'itchyny/lightline.vim'
Plug 'itchyny/vim-gitbranch'
Plug 'airblade/vim-gitgutter'
Plug 'mengelbrecht/lightline-bufferline'
Plug 'raimondi/delimitmate'
"Plug 'vim-airline/vim-airline'

"color schemes
Plug 'arcticicestudio/nord-vim'
Plug 'kaicataldo/material.vim', { 'branch': 'main' }
"Plug 'vim-airline/vim-airline-themes'

call plug#end()
```

2. ligthline.vim config.

```vimrc
"------------------------------------------------------------------------
  " lightline.vim config.
"------------------------------------------------------------------------

"if the statusline/buffuerline does not render propoerly add this line:
set laststatus=2
set showtabline=2

"define options
let g:lightline = {}
let g:lightline.colorscheme = 'nord'
let g:lightline.active = {
            \ 'left': [
            \   ['mode'],
            \   ['gitbranch', 'githunks', 'readonly'],
            \   ['filename']
            \ ],
            \ 'right': [
            \   ['whitespace', 'syntastic', 'lineinfo'],
            \   ['percent'],
            \   ['fileformat', 'fileencoding', 'filetype']
            \ ]
            \ }
let g:lightline.component_function = {
            \ 'githunks': 'LightlineGitGutter',
            \ 'gitbranch': 'LightlineGitbranch',
            \ 'readonly1': 'LightlineReadonly',
            \ 'filename': 'LightlineFilename'
            \ }
let g:lightline.component_expand = {
            \ 'syntastic': 'SyntasticStatuslineFlag',
            \ 'whitespace': 'lightline#whitespace#check'
            \ }
let g:lightline.component_type = {
            \ 'syntastic': 'error',
            \ 'whitespace': 'warning'
            \ }
let g:lightline.component = {
            \ 'modified': '%{&modified?"+":""}',
            \ 'readonly': '%{&readonly?"⊝":""}'
            \ }
let g:lightline.separator = {
            \ 'left': "\ue0b0",
            \ 'right': "\ue0b2"
            \ }
let g:lightline.subseparator = {
            \ 'left': "\ue0b1",
            \ 'right': "\ue0b3"
            \ }

"lightline component functions
function! LightlineGitGutter()
  if !get(g:, 'gitgutter_enabled', 0) || empty(FugitiveHead())
    return ''
  endif
  let [ l:added, l:modified, l:removed ] = GitGutterGetHunkSummary()
  return printf('+%d ~%d -%d', l:added, l:modified, l:removed)
endfunction

function! LightlineGitbranch()
  if exists("*gitbranch#name")
    let branch = gitbranch#name()
    return branch !=# '' ? ' '.branch : ''
  endif
  return ''
endfunction

function! LightlineReadonly()
  return &readonly ? '' : ''
endfunction

function! LightlineFilename()
  let filename = expand('%:t') !=# '' ? expand('%:t') : '[No Name]'
  let modified = &modified ? ' +' : ''
  return filename . modified
endfunction
```

3. Syntastic config.

```vimrc
"------------------------------------------------------------------------
  " Syntastic config.
"------------------------------------------------------------------------

set statusline+=%#warningmsg#
set statusline+=%{SyntasticStatuslineFlag()}
set statusline+=%*

let g:syntastic_always_populate_loc_list = 1
let g:syntastic_check_on_open = 1
let g:syntastic_check_on_wq = 0

"mark syntax errors with :signs
let g:syntastic_enable_signs = 1

"automatically jump to the error when saving the file
let g:syntastic_auto_jump = 1

"show the error list automatically
let g:syntastic_auto_loc_list = 1

"show warnings
let g:syntastic_quite_message = 0

"mode option
let g:syntastic_mode_map = { 'mode': 'passive',
      \                      'active_filetypes': ['md', 'c', 'cpp'] }

"lint engine options
let g:syntastic_markdown_mdl_exec = "markdownlint"
let g:syntastic_markdown_mdl_args = ""
let g:syntastic_markdown_checkers = ['proselint']
let g:syntastic_json_checkers = ['jsonlint']
let g:syntastic_sql_checkers = ['sqlint']
```

**Comment by itchyny** [comment]
For syntastic plugin, refer to `:h syntastic-loclist-callback` and `call lightline#update()` on the check hook. Also, the whitespace plugin looks working for me, when `call lightline#update()`. The expanded components are evaluated on this function, normally on switching tabs or windows. This is an expected behavior.

**Comment by ondiversion** [comment]
@itchyny thank you for your advice. since i can barely copying others' .vimrc, it was quite challenging to follow your suggested solution.. I blame myself for the trouble. honestly I gave up on syntastic, since I discovered [lightline.ale](https://github.com/maximbaz/lightline-ale). installing ale and lightline ale works out-of-the-box! and in case of the lightline whitespace plugin, I wrote lightline configuration in .vimrc from the scratch, again. now it works perfectly along with lightline ale. i failed to figure out what prevented lightline whitespace plugin from proper functioning. 

anyway, thank you for your awseome plugin and expandability which comes with your plugin.
i am really appreciate your commitment.
cheers!
